tokio/src/runtime/time/wheel/cancellation_queue.rs

```diff
@@ -0,0 +1,257 @@
+
+fn spin_loop() {
+    #[cfg(loom)]
+    crate::loom::thread::yield_now();
+
+    #[cfg(not(loom))]
+    std::hint::spin_loop();
+}
+
+#[derive(Debug)]
+struct Inner {
+    head: UnsafeCell<NonNull<Entry>>,
+    tail: AtomicPtr<Entry>,
+    stub: AtomicPtr<Entry>,
+}
+
+unsafe impl Send for Inner {}
+unsafe impl Sync for Inner {}
+
+impl Drop for Inner {
+    fn drop(&mut self) {
+        unsafe {
+            let _ = self.take_all();
+        }
+        // Drop the stub pointer
+        let stub = NonNull::new(self.stub.load(SeqCst)).unwrap();
+        drop_stub(stub);
+    }
+}
+
+impl Inner {
+    pub(crate) fn new() -> Self {
+        let stub = new_stub();
+
+        Self {
+            head: UnsafeCell::new(NonNull::new(stub.as_ptr()).unwrap()),
+            tail: AtomicPtr::new(stub.as_ptr()),
+            stub: AtomicPtr::new(stub.as_ptr()),
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Violating any of the following constraints can lead to
+    /// undefined behavior:
+    ///
+    /// - `hdl` must not be in any queue.
+    unsafe fn push(&self, hdl: EntryHandle) {
+        // Since all items in the queue must be alive until they are removed,
+        // so we should not decrease the reference count.
+        let node = ManuallyDrop::new(hdl.into_entry());
+
+        let next = node.cancel_pointer();
+        next.store(null_mut(), SeqCst);
+
+        let old_tail = self.tail.swap(Arc::as_ptr(&node).cast_mut(), SeqCst);
+        old_tail
+            .as_ref()
+            .expect("tail pointer should never be null")
+            .cancel_pointer()
+            .store(Arc::as_ptr(&node).cast_mut(), SeqCst);
+    }
+
+    /// # Safety
+    ///
+    /// Violating any of the following constraints can lead to
+    /// undefined behavior:
+    ///
+    /// - This method must not be called concurrently.
+    unsafe fn take_all(&self) -> impl Iterator<Item = EntryHandle> {
+        // TODO: Using `Option` for both head and tail is a bad design,
+        // imagine a case where the head is None, but the tail is Some,
+        // which is very confusing.
+        struct Iter {
+            head: Option<NonNull<Entry>>,
+            tail: Option<NonNull<Entry>>,
+        }
+
+        impl Drop for Iter {
+            fn drop(&mut self) {
+                for hdl in self {
+                    drop(hdl)
+                }
+            }
+        }
+
+        impl Iterator for Iter {
+            type Item = EntryHandle;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                match self.head {
+                    Some(head) => unsafe {
+                        let atomic_next = head.as_ref().cancel_pointer();
+                        let mut next = atomic_next.load(SeqCst);
+                        while head != self.tail.unwrap() && next.is_null() {
+                            spin_loop();
+                            next = atomic_next.load(SeqCst);
+                        }
```

Comment:
Spinloops are generally pretty problematic performance-wise. Can we just protect the linked list by a mtuex instead of using atomics?

Reply:
I think the 'window of spin loop' is extremely small, I drew a diagram to explain it.

<img width="11402" height="8856" alt="queue" src="https://github.com/user-attachments/assets/9a00116c-93aa-4389-a254-aaea7c0f4088" />